python/ray/client_builder.py

```diff
@@ -29,12 +29,18 @@ class ClientContext:
     ray_commit: str
     protocol_version: Optional[str]
     _num_clients: int
+    ray_cli: Optional[ray.util.client.RayAPIStub]
 
     def __enter__(self) -> "ClientContext":
+        if self.ray_cli is not None:
+            self.ray_cli = ray.util.client.ray.set_context(self.ray_cli)
         return self
 
     def __exit__(self, *exc) -> None:
-        self.disconnect()
+        if self.ray_cli is None or ray.util.client.is_main_cli():
+            self.disconnect()
```

Comment:
Hmm, we might need an explicit disconnect here to actually disconnect from the cluster (or we can just wait until the python program exits).


I definitely agree that removing disconnect here is critical because we want to do:
```
client_context = ray.client().connect()
with client_context:
   # do work

with client_context:
   # do some more work
```

Reply:
I won't remove it right now, since it has too many updates. And it also requires API change. I'll try to make them compatible for now:

- single client mode, it will disconnect automatically
- multi client mode, it wont disconnect.

Reply:
Let me bring this up in api-changes. We definitely don't want two contradicting use cases for with ray.init().